File: src/leetcode/m129/SumRootToLeafNumbers.java
/*
 * Copyright (c) 2024 THGEEK.COM. All rights reserved.
 */

package leetcode.m129;

import leetcode.TreeNode;

import java.util.ArrayList;
import java.util.List;

/**
 * LeetCode 129
 *
 * @author Tao Hong
 * @version 1.0
 * @since 2024/10/08 08:20
 */
public class SumRootToLeafNumbers {
    private final List<List<Integer>> paths = new ArrayList<>();

    public int sumNumbers(TreeNode root) {
        addPath(root, new ArrayList<>());
        int res = 0;
        for (List<Integer> path : paths) {
            int digit = 0;
            for (int i = path.size() - 1; i >= 0; i--) {
                res += path.get(i) * (int) Math.pow(10, digit++);
            }
        }
        return res;
    }

    private void addPath(TreeNode root, List<Integer> path) {
        if (root.left == null && root.right == null) {
            path.add(root.val);
            paths.add(path);
            return;
        }
        path.add(root.val);
        List<Integer> right = new ArrayList<>(path);
        if (root.left != null) {
            addPath(root.left, path);
        }
        if (root.right != null) {
            addPath(root.right, right);
        }
    }
}

File: src/leetcode/m129/SumRootToLeafNumbers.go
package m129

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// LeetCode 129
func sumNumbers(root *TreeNode) int {
	return sumNumberSlice(root, 0)
}

func sumNumberSlice(node *TreeNode, sum int) int {
	if node == nil {
		return sum
	}
	sum = sum*10 + node.Val
	if node.Left == nil && node.Right == nil {
		return sum
	}
	if node.Left == nil {
		return sumNumberSlice(node.Right, sum)
	}
	if node.Right == nil {
		return sumNumberSlice(node.Left, sum)
	}
	leftSum := sumNumberSlice(node.Left, sum)
	rightSum := sumNumberSlice(node.Right, sum)
	return leftSum + rightSum
}
